File: same_tree.py
# 100. Same Tree

"""
given the roots of two binary trees 'p' and 'q', write a function to check if
they are the same or not. two binary trees are considered the same if they
are structurally identical, and the nodes have the same value.
"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if p is None and q is None:
            return True
        if p is None or q is None:
            return False
        if p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        return False

File: same_tree.c
// 100. Same Tree
#include "leetcode.h"

/*
 * given the roots of two binary trees 'p' and 'q', write a function to check if
 * they are the same or not. two binary trees are considered the same if they
 * are structurally identical, and the nodes have the same value.
 */

struct TreeNode {
  int val;
  struct TreeNode *left, *right;
};

bool isSameTree(struct TreeNode *p, struct TreeNode *q) {
  if (!p && !q)
    return 1;
  else if (!p || !q)
    return 0;
  return p->val == q->val
             ? isSameTree(p->left, q->left) && isSameTree(p->right, q->right)
             : 0;
}
